### PR TITLE
self-hosted runner: always use latest Git for Windows and GitHub Actions version (plus some PowerShell cleanups)

### PR DESCRIPTION
**Pull request by rstolpe**
**GitHub git-windows and Action runner**
- Added check as was asked for both ActionRunner and GitHub git-windows
- Now all data for GitHub git-windows and ActionRunner are stored in the objects $GitHubAction and $GitHubGit
- Used invoke-restmethod to eliminate overhead and it makes it easier to go over to PSCore later on. Also I added the accept and API version in the header. With this you don't need to convert it from json etc. it's easier.

**Ensure that Action runner service was created**
Added so it will look if it has been created 3 times 3 seconds apart before it write error as sometime it can take time for the service to be created.

**Stop Service, actions.runner.***
- As you was using * I'm guessing that it can be many services so I did add an foreach loop on them.
- Inside the foreach it will make sure that each service has been stopped before continuing, it will look 5 times with 5 seconds apart.

**Param**
- Moved text to helpmessage
- Added validation
- Changed StopService to bool so now you need to write -StopService $false if you don't want to stop the service. Default it's $true as it was from the start.

**Error handling**
- Replaced throw with write-error and exit 1 as I'm guessing that you want to exit the script on this errors.

**Other**
- Removed $ProgressPreferences where it was not needed.

Got some more things that I can do later on when I have the time, I have tested it locally at the code should work but as I can't test with ActionRunner I'm advising you to test it also.

### CODE REVIEW

**Comment by rstolpe** [comment]
Look in the conversation for this PR it contains more information, but in short I have:
- improved collection from GitHub git-windows and action runner
- Removed $ProgressPreferences where it was not needed.
- Improved param
- Improved error handeling
- Improved verification that services has been stopped or created

**Comment by rstolpe** [comment]
@dscho forget to req review on this one but when you have the time, take a look at it :)

File: azure-self-hosted-runners/post-deployment-script.ps1
Comment (dscho) [inline azure-self-hosted-runners/post-deployment-script.ps1:55]: I learned something new today: I had not known that `$Matches` is set automatically by `-match`. Thank you!

File: azure-self-hosted-runners/post-deployment-script.ps1
Comment (dscho) [inline azure-self-hosted-runners/post-deployment-script.ps1:191]: This is really interesting information. Since the post-deployment script is run on Azure, and [the GitHub workflow that kicks off that deployment](https://github.com/git-for-windows/git-for-windows-automation/actions/workflows/create-azure-self-hosted-runners.yml) has no connection to the actual deployment, I wonder whether there is any way we could manually mirror it into the GitHub workflow run's logs... @dennisameling any ideas?

Comment (dscho) [reply azure-self-hosted-runners/post-deployment-script.ps1:191]: Oooh, ooh, I think I found something: https://learn.microsoft.com/en-us/powershell/module/az.resources/get-azdeploymentscriptlog?view=azps-9.3.0

File: azure-self-hosted-runners/post-deployment-script.ps1
Comment (dscho) [inline azure-self-hosted-runners/post-deployment-script.ps1:9]: Should this be `-like` instead? We expect this to be a HTTPS URL, right?

File: azure-self-hosted-runners/post-deployment-script.ps1
Comment (dscho) [inline azure-self-hosted-runners/post-deployment-script.ps1:17]: I had actually tried this, and that did _not_ get picked up because we pass a string from the GitHub workflow. So I would suggest to drop this change.

Comment (dscho) [reply azure-self-hosted-runners/post-deployment-script.ps1:17]: Oh, wait... That was not the `post-deployment` script variable, but the workflow input, and there a `boolean` really would not work.

**Review by dscho — COMMENTED** [review]
This is wonderful!

I would like to wait for @dennisameling's review before merging this because they are much more familiar with this stuff than I am (I basically only cargo-culted my work on this so far).

Apart from the one change to a `[bool]` (which I am not so sure about), I only have two minor suggestions (but I won't insist on them): in the Git project, we usually craft commit messages [with a lot more information](https://github.blog/2022-06-30-write-better-commits-build-better-projects/). Oh, and every time I read "GitHub" with a lower-case "h" my brain mis-pronounces that word; PowerShell might be case-insensitive, but I am not 😁

File: azure-self-hosted-runners/post-deployment-script.ps1
Comment (rstolpe) [reply azure-self-hosted-runners/post-deployment-script.ps1:9]: That's correct I'll change it

File: azure-self-hosted-runners/post-deployment-script.ps1
Comment (rstolpe) [reply azure-self-hosted-runners/post-deployment-script.ps1:17]: Ok I understand, I'll change it to string again. Would a switch work instead? Then you just need to write -Stopservice and it's true so if you don't include the -Stopservice in your command it's false.
I'm kind of alergic against strings when you use true or fasle :P But I'm changing it.

File: azure-self-hosted-runners/post-deployment-script.ps1
Comment (rstolpe) [reply azure-self-hosted-runners/post-deployment-script.ps1:55]: :)

File: azure-self-hosted-runners/post-deployment-script.ps1
Comment (rstolpe) [reply azure-self-hosted-runners/post-deployment-script.ps1:191]: I'm not sure what you mean? Do you want to log information from the script? If so, that's possible. I can even make somethign that can send you a email if you want.

**Comment by rstolpe** [comment]
@dscho i know I need to train on my comments on the commits. Same at work. I added some better ones for mistakes I was correcting

I wanted to use switch instead of bool but as I don't know how you execute the script I don't know if switch was good to use or not.

String for true or false is a big no ;) but I did revert it back

I'll adjust the script a little more during the end of next week I have some more adjustments that I want to add to make it even better.

Sorry that my last commits was unsigned I did not have any pgp cert for GitKraken but now I have fixed that too, thought it would use my ssl cert that vs and vscode are using.

**Comment by dscho** [comment]
Oh, you don't need to bother with GPG-signing... I was thinking more about giving context in the commit messages, intentions and additional information that's not obvious from reading the diff.

About using `[bool]`: if this works by providing the command-line parameter as a _string_, then by all means, let's use the correct data type.

**Comment by dennisameling** [comment]
> I don't know how you execute the script I don't know if switch was good to use or not.

We use Microsoft's Azure Custom Script Extension for it. [Here's how it's invoked](https://github.com/git-for-windows/git-for-windows-automation/blob/200fdc9be4b4af6dc8751f2e18202eeab9ff5253/azure-self-hosted-runners/azure-arm-template.json#L265). So I guess that something like this happens behind the scenes:

```
C:\PATH_TO_SCRIPT\post-deployment-script.ps1 -GitHubActionsRunnerToken ABC123 -GithubActionsRunnerRegistrationUrl https://example.com -GithubActionsRunnerName ABC123 -StopService false -GitHubActionsRunnerPath D:\a
```

I'll test your changes on a fresh VM later today :) thanks for your work on this!

**Comment by rstolpe** [comment]
> > I don't know how you execute the script I don't know if switch was good to use or not.
> 
> 
> 
> We use Microsoft's [Azure Custom Script Extension](https://learn.microsoft.com/en-us/azure/virtual-machines/extensions/custom-script-windows) for it. [Here's how it's invoked](https://github.com/git-for-windows/git-for-windows-automation/blob/200fdc9be4b4af6dc8751f2e18202eeab9ff5253/azure-self-hosted-runners/azure-arm-template.json#L265). So I guess that something like this happens behind the scenes:
> 
> 
> 
> ```
> 
> powershell -ExecutionPolicy Unrestricted -File C:\PATH_TO_SCRIPT\post-deployment-script.ps1 -GitHubActionsRunnerToken ABC123 -GithubActionsRunnerRegistrationUrl https://example.com -GithubActionsRunnerName ABC123 -StopService false -GitHubActionsRunnerPath D:\a
> 
> ```
> 
> 
> 
> I'll test your changes on a fresh VM later today :) thanks for your work on this!

I did return it to string but as it look of it I can improve it more now when I know how you execute it.

But test this things out first and after that I during next week improve it more :)

**Review by dennisameling — APPROVED** [review]
Just ran the script on a fresh VM. Everything looks great. Also really like how you approached this in PowerShell. Thanks @rstolpe! 🚀 

<details>
  <summary>Script log</summary>

  <pre>
Starting post-deployment script.
Enabling Windows Developer Mode...
Enabled Windows developer mode.
Adding Microsoft Defender Exclusions...
Finished adding Microsoft Defender Exclusions.
Downloading Git for Windows...
Installing Git for Windows...
Finished installing Git for Windows.
Downloading GitHub Actions runner...
Installing GitHub Actions runner 2.300.2 as a Windows service with labels self-hosted,Windows,ARM64...
Configuring the runner to shut down automatically after running
Configuring the runner

--------------------------------------------------------------------------------
|        ____ _ _   _   _       _          _        _   _                      |
|       / ___(_) |_| | | |_   _| |__      / \   ___| |_(_) ___  _ __  ___      |
|      | |  _| | __| |_| | | | | '_ \    / _ \ / __| __| |/ _ \| '_ \/ __|     |
|      | |_| | | |_|  _  | |_| | |_) |  / ___ \ (__| |_| | (_) | | | \__ \     |
|       \____|_|\__|_| |_|\__,_|_.__/  /_/   \_\___|\__|_|\___/|_| |_|___/     |
|                                                                              |
|                       Self-hosted runner registration                        |
|                                                                              |
--------------------------------------------------------------------------------

# Authentication


√ Connected to GitHub

# Runner Registration




√ Runner successfully added
√ Runner connection is good

# Runner settings


√ Settings Saved.

Granting file permissions to 'NT AUTHORITY\NETWORK SERVICE'.
Service actions.runner.dennisameling-git-for-windows-automation.arm64-test-vm-dennis successfully installed
Service actions.runner.dennisameling-git-for-windows-automation.arm64-test-vm-dennis successfully set recovery option
Service actions.runner.dennisameling-git-for-windows-automation.arm64-test-vm-dennis successfully set to delayed auto start
Service actions.runner.dennisameling-git-for-windows-automation.arm64-test-vm-dennis successfully configured
Waiting for service to start...
Service actions.runner.dennisameling-git-for-windows-automation.arm64-test-vm-dennis started successfully
Finished installing GitHub Actions runner.
</pre>
</details>

**Comment by rstolpe** [comment]
@dennisameling ill optimize it some more at the end of next week but until that I do think this will work for you.

It's nice to contribute to other projects.

Just DM me if you have some other things you want automated or similar :)

File: azure-self-hosted-runners/post-deployment-script.ps1
Comment (dennisameling) [reply azure-self-hosted-runners/post-deployment-script.ps1:192]: > Oooh, ooh, I think I found something: https://learn.microsoft.com/en-us/powershell/module/az.resources/get-azdeploymentscriptlog?view=azps-9.3.0

Nice try, but [Deployment Scripts](https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/deployment-script-template) are something else than the [Custom Script Extension](https://learn.microsoft.com/en-us/azure/virtual-machines/extensions/custom-script-windows) 😉 

Let's follow up in [this PR](https://github.com/git-for-windows/git-for-windows-automation/pull/21) for getting the post-deployment script output 👍🏼

**Comment by dennisameling** [comment]
@dscho could you do a final review? I guess you'll want some more descriptive commit messages, right?

**Comment by dscho** [comment]
> @dscho could you do a final review? I guess you'll want some more descriptive commit messages, right?

Maybe next time the commit messages could follow the [Intent, Context, Implementation & Justification format](https://github.blog/2022-06-30-write-better-commits-build-better-projects/).

But I don't want to hold up this contribution any further 😄

**Review by dscho — APPROVED** [review]

**Pull request merged**
